% -*- mode: noweb; noweb-default-code-mode: R-mode; -*-
\documentclass{article}

%headers and footers
\usepackage{fancyhdr}
\setlength{\headheight}{15pt}
\pagestyle{fancyplain}
\lhead{\fancyplain{}{\thepage}}
\chead{}
\rhead{\fancyplain{}{\bfseries miRNA Reanalyis}}
\rfoot{$Rev: $}
\cfoot{}
\lfoot{\includegraphics[width=0.1\textwidth]{./bicLogo.png}\\[1cm]}


%hyperlink setup
\usepackage{hyperref}
\usepackage{xcolor}
\definecolor{dark-red}{rgb}{0.4,0.15,0.15}
\definecolor{dark-blue}{rgb}{0.15,0.15,0.4}
\definecolor{medium-blue}{rgb}{0,0,0.5}
\hypersetup{
    colorlinks, linkcolor={dark-red},
    citecolor={dark-blue}, urlcolor={medium-blue}
}


\usepackage{longtable}
\usepackage{rotating}


\begin{document}
\newcommand{\COT}{C$_{0}$t }
\title{A Reanalysis of miRNA differential expression, editing, and novelty}
\author{Jeremy Leipzig}
\maketitle
\tableofcontents
\pagebreak
<<setup,echo=FALSE>>=
source("config.R")
library(DESeq)
library(ggplot2)
library(xtable)
library(Biostrings)

@
\section{Introduction}
miRNA counts were generated by Johnathan Schug using an Eland
alignment of miRNAs to mature miRNA sequences in mirbase and the
genome.


This analysis was revisited with the following objectives:
\begin{enumerate}
  \item Reanalyze existing miRNA count data using an existing RNA-Seq package with an explicitly documented procedure for comparing samples without replicates
  \item Use a non-heuristic aligner to reassess the existing counts
    and try the RNA-Seq analysis on these counts
  \item Explore the possibility of miRNA editing
  \item Explore potential novel miRNAs in the dataset and their
    predicted targets
\end{enumerate}

\section{Samples}
<<loadSeqs,echo=F,results=hide>>=


runs<-data.frame(run=c('FGC0042S1','FGC0031S8','FGC0036S1','FGC0036S2'),desc=c('Normal retina','WERI with barcode GTCT (Retinoblastoma)','Y-79 without barcode (Retinoblastoma)','Y-79 + WERI with barcode (Retinoblastoma)'),size=c(4787704,5821153,5877018,5897869))
@
Four Solexa runs were conducted:
\begin{enumerate}
   \item \Sexpr{paste(runs[1,1],':',runs[1,2],' ',format(runs[1,3],big.mark=",",scientific=F),' seqs',sep="")}
   \item  \Sexpr{paste(runs[2,1],':',runs[2,2],' ',format(runs[2,3],big.mark=",",scientific=F),' seqs',sep="")}
   \item \Sexpr{paste(runs[3,1],':',runs[3,2],' ',format(runs[3,3],big.mark=",",scientific=F),' seqs',sep="")}
   \item \Sexpr{paste(runs[4,1],':',runs[4,2],' ',format(runs[4,3],big.mark=",",scientific=F),' seqs',sep="")}
\end{enumerate}


\section{Reanalysis of existing count data}
\subsection{Schug analysis and confusion}
In the absence of biological replicates it is presumably impossible to
distinguish the real effect of an independent variable and the vagarities
of individual variability. Assumptions can be made
leveraging what data is available within each sample, as well as
empirical evidence from other experiments.

The existing calculation of probabilities from Johnathan
Schug were based on fold change and some unknown calculation of
noise garnered from a microarray analysis approach.

A workflow document was given although the source code was not included.
A summary report was not provided.

Four result spreadsheets were discovered in:
\begin{verbatim}
./RB_miRNA_raw_data/559-Ganguly-Chao-Solexa/Report/
\end{verbatim}

\begin{enumerate}
  \item ba-OneLaneWy.xls
  \item ba-TwoLaneWy.xls
  \item ba-Weri.xls
  \item ba-Y79.xls
\end{enumerate}

From the correspondence with Arupa Ganguly it was originally thought
TwoLaneWy was a comparison between Normal and 36S2Weri but this file
is only found in the 559 experimental directory. It appears the
lane comprising the normal retina was not sequenced until experiment 589.

The confusion likely arose because files produced in 589 have the same
names (Weri,Y79,TwoLaneWy,OneLaneWy). This coincidence is documented
in the 589 workflow document.

Using the 559 key:
\begin{verbatim}
1 FGC0031-8_GTCT
2 FGC0036-1_NOBC
3 FGC0036-2_GTCT
4 FGC0036-2_NOBC
\end{verbatim}
and workflow
\begin{verbatim}
ControlCond=1 TestCond=3 Name=Weri ;
ControlCond=2 TestCond=4 Name=Y79 ;
ControlCond=1 TestCond=2 Name=TwoLaneWy ;
ControlCond=3 TestCond=4 Name=OneLaneWy ;
\end{verbatim}xs
We see TwoLaneWy (1 vs 2) was likely a comparison of 31S8 WERI vs. 36S1 Y-79.

with following columns
\begin{enumerate}
  \item Gene - human mature miRNA
  \item ControlN and TestN - presumably non-normalized counts
  \item FoldChange and abs(FC), unknown equation presumably based on raw counts
  \item M (typically a red/green intensity ratio) definitely calculated as log$_2$ fold change
  \item A (average intensity) unknown calculation (typically (log$_2$test+log$_2$ctrl)/2)
  \item ControlP and TestP - some fraction of respective sample totals
  \item Ln(Pv) and Statistic and Heavy - unknown
\end{enumerate}

(Dr. Schug was unable or unwilling to respond to my emails
concerning this analysis).


\subsection{DESeq}
A reanalysis of the TwoLaneWy count data was performed using the
R-based RNA-Seq analysis package DESeq \cite{Anders} in order validate
and reassess the significance calls from that spreadsheet. When
spreadsheets containing Schug's comparisons of Normal vs WERI/Y-79
data are located, they can be reassessed as well.

To deal with a lack of replicates, DESeq uses variability of
normalized gene counts with similar means, under the assumption that
most genes are not differentially expressed
(\href{http://seqanswers.com/forums/showthread.php?t=7325}{seqanswers post}). This might not be the best assumption for miRNAs.


\begin{quote}
uses the negative binomial distribution, with variance and mean linked by local regression, to model the null distribution of the count data. The method controls type-I error and provides good detection power. "
\end{quote}

Due to the lack of replicates in this data set, a "blind" method of variance estimation must be used. This approach is described by Simon Anders:
\begin{quote}
[experiments without replicates] are often undertaken, especially in HTS, and the DESeq package can deal with them, even though the soundness of the results may depend very much on the circumstances.
Our primary assumption is still that the mean is a good predictor for the variance. Hence, if a number of genes with similar expression level are compared between replicates, we expect that their variation is of comparable magnitude. Once we accept this assumption, we may argue as follows: Given two samples from different conditions and a number of genes with comparable expression levels, of which we expect only a minority to be influenced by the condition, we may take the variance estimated from comparing their count rates across conditions as ersatz for a proper estimate of the variance across replicates. After all, we assume most genes to behave the same within replicates as across conditions, and hence, the estimated variance should not change too much due to the influence of the hopefully few differentially expressed genes. Furthermore, the differentially expressed genes will only cause the variance estimate to be too high, so that the test will err to the side of being too conservative, i.e., we only lose power.
\end{quote}

<<deseqBaTwoWay,echo=F>>=
baTwoWay<-read.csv(paste(gangulyRoot,"RB_miRNA_raw_data/559-Ganguly-Chao-Solexa/Report/ba-TwoLaneWy.csv",sep="/"))
samp1<-"31S8_WERI"
samp2<-"36S1_Y79"
numTrueRows<-length(which(baTwoWay$Significant==TRUE))


countsTable<-baTwoWay[,c("ControlN","TestN")]

row.names(countsTable)<-baTwoWay$Gene
conds<-as.factor(c(samp1,samp2))
cds<-newCountDataSet(countsTable,conds)
cds<-estimateSizeFactors(cds)
cds<-estimateVarianceFunctions(cds,method="blind")
res<-nbinomTest(cds,samp1,samp2)
write.table(res,file=paste(samp1,"_",samp2,".txt",sep=""))
myPval<-.05
@
\pagebreak
\subsubsection{BaTwoLaneWy}
Data from the BaTwoLaneWy.xls spreadsheet, reportedly a comparison of FGC0042S1(Normal retina) with the WERI barcode sequences from FGC0036S2 aligned against \Sexpr{nrow(baTwoWay)} mature miRNAs, of which \Sexpr{numTrueRows} were considered significantly enriched or depleted.

<<dePval,results=tex,echo=F>>=

myXtableOne<-xtable(subset(res,pval<myPval,1:8),type=tex,caption=paste("Significant (P<",myPval,") differential expression before the correction for multiple testing"))
print(myXtableOne,include.rownames=FALSE, size="\\footnotesize")
@

After correcting for multiple testing (Benjamini-Hochberg procedure),
\Sexpr{nrow(subset(res,padj<.05))} miRNAs were differential
expressed. I beleive these are both retinoblastoma samples and so the
lack of differential expression not discouraging.

If the 589 version of baTwoWay.xls is located we can easily revisit that analysis.


\section{Realignment using Vmatch}

Eland is a first generation short read mapper with numerous
limitations.

Due to the small libraries of known miRNAs, it was possible to use a
non-heuristic general alignment tool, Vmatch, to gain total confidence
and flexibility with the distinct 11-26nt sequences collated by Schug.

\subsection{Trimmed files}
Schug had already quality trimmed, adaptor-masked, poly-A trimmed and debarcoded the
miRNA sequences from each run. In addition, sequences were split by
sequence length, presumably to focus in on true miRNA candidates and possibly
to aid Eland, which in the early days of Solexa did not accept
variable length reads.

This preprocessing comprised a decent amount of labor, and so lacking the original
adapter sequences and other minutae, so it was chosen to accept these trimmed sequences rather than start from scratch.

<<loadTrimmed,echo=F>>=
seqs<-read.table("allSeqs.txt",col.names="filePath",stringsAsFactors=FALSE)

#breakup set identifiers based on filepath, we are lucky the charlengths are all the same
seqs$seqLength<-as.integer(substr(seqs$filePath,nchar(seqs$filePath)-4,nchar(seqs$filePath)-3))
seqs$expLane<-substr(seqs$filePath,nchar(seqs$filePath)-19,nchar(seqs$filePath)-11)
seqs$barcode<-substr(seqs$filePath,nchar(seqs$filePath)-9,nchar(seqs$filePath)-6)

seqs[which(seqs$expLane=='FGC0042-1'),"sample"]<-"normal"
seqs[which(seqs$barcode=='GTCT'),"sample"]<-"WERI"
seqs[which((seqs$expLane=='FGC0036-1' | seqs$expLane=='FGC0036-2') & seqs$barcode=='NOBC'),"sample"]<-'Y-79'

#get number of sequences
seqs$numSeqs<-laply(.data=paste(gangulyRoot,seqs$filePath,sep="/"),function(x){length(read.DNAStringSet(x))})
write.table(seqs,file="schugSeqsTable.txt")
@

It was determined that no sequence within a lane/sample was repeated -
i.e. these are distinct sequences.

A breakdown of these by lane and sample

<<seqTab,results=tex,echo=F>>=
seqTabulation<-ddply(seqs,c("sample","expLane"),summarize,distinctSeqs=sum(numSeqs))
myXtableTwo<-xtable(seqTabulation,type=tex,caption=paste("Distinct sequences by sample/lane"))
print(myXtableTwo,include.rownames=FALSE, size="\\footnotesize")
@



\begin{thebibliography}{100}	% start the bibliography

\small				% put the bibliography in a small font

\bibitem{Anders} Anders, S. and  Huber, W. Differential expression analysis for sequence count data. Nature Precedings
http://dx.doi.org/10.1038/npre.2010.4282.2.

\end{thebibliography}		% end the bibliography


\end{document}
