% -*- mode: noweb; noweb-default-code-mode: R-mode; -*-
\documentclass{article}

%headers and footers
\usepackage{fancyhdr}
\setlength{\headheight}{15pt}
\pagestyle{fancyplain}
\lhead{\fancyplain{}{\thepage}}
\chead{}
\rhead{\fancyplain{}{\bfseries miRNA Reanalyis}}
\rfoot{$Rev$}
\cfoot{}
\lfoot{\includegraphics[width=0.1\textwidth]{./bicLogo.png}\\[1cm]}


%hyperlink setup
\usepackage{hyperref}
\usepackage{xcolor}
\definecolor{dark-red}{rgb}{0.4,0.15,0.15}
\definecolor{dark-blue}{rgb}{0.15,0.15,0.4}
\definecolor{medium-blue}{rgb}{0,0,0.5}
\hypersetup{
    colorlinks, linkcolor={dark-red},
    citecolor={dark-blue}, urlcolor={medium-blue}
}


\usepackage{longtable}
\usepackage{rotating}

%underscores in variable names cause problems
%\usepackage{underscore}

\begin{document}
\newcommand{\COT}{C$_{0}$t }
\title{A Reanalysis of miRNA differential expression, editing, and novelty}
\author{Jeremy Leipzig}
\maketitle
\tableofcontents
\pagebreak
<<setup,echo=FALSE>>=
source("config.R")
library(DESeq)
library(ggplot2)
library(xtable)
library(Biostrings)

if(!exists("samp1") | !exists("sizeName1") | !exists("useLibs")){stop('Missing startup info')}
@
\section{Introduction}
miRNA counts were generated by Johnathan Schug using an Eland
alignment of miRNAs to mature miRNA sequences in mirbase and the
genome.


This analysis was revisited with the following objectives:
\begin{enumerate}
  \item Reanalyze existing miRNA count data using an existing RNA-Seq package with an explicitly documented procedure for comparing samples without replicates
  \item Use a non-heuristic aligner to reassess the existing counts
    and try the RNA-Seq analysis on these counts
  \item Explore the possibility of miRNA editing
  \item Explore potential novel miRNAs in the dataset and their
    predicted targets
\end{enumerate}

\section{Samples}
<<loadSeqs,echo=F,results=hide>>=


runs<-data.frame(run=c('FGC0042S1','FGC0031S8','FGC0036S1','FGC0036S2'),desc=c('Normal retina','WERI with barcode GTCT (Retinoblastoma)','Y-79 without barcode (Retinoblastoma)','Y-79 + WERI with barcode (Retinoblastoma)'),size=c(4787704,5821153,5877018,5897869))
@
Four Solexa runs were conducted:
\begin{enumerate}
   \item \Sexpr{paste(runs[1,1],':',runs[1,2],' ',format(runs[1,3],big.mark=",",scientific=F),' seqs',sep="")}
   \item  \Sexpr{paste(runs[2,1],':',runs[2,2],' ',format(runs[2,3],big.mark=",",scientific=F),' seqs',sep="")}
   \item \Sexpr{paste(runs[3,1],':',runs[3,2],' ',format(runs[3,3],big.mark=",",scientific=F),' seqs',sep="")}
   \item \Sexpr{paste(runs[4,1],':',runs[4,2],' ',format(runs[4,3],big.mark=",",scientific=F),' seqs',sep="")}
\end{enumerate}


\section{Reanalysis of existing count data}
\subsection{Schug analysis}
In the absence of biological replicates it is presumably impossible to
distinguish the real effect of an independent variable and the vagarities
of individual variability. Assumptions can be made
leveraging what data is available within each sample, as well as
empirical evidence from other experiments.

The existing calculation of probabilities from Johnathan
Schug were based on fold change and some unknown calculation of
noise garnered from a microarray analysis approach.

A workflow document was given although the source code was not included.

Four result spreadsheets were discovered in:
\begin{verbatim}
./RB_miRNA_raw_data/559-Ganguly-Chao-Solexa/Report/
\end{verbatim}

\begin{enumerate}
  \item ba-OneLaneWy.xls
  \item ba-TwoLaneWy.xls
  \item ba-Weri.xls
  \item ba-Y79.xls
\end{enumerate}

It was originally thought TwoLaneWy was a comparison between Normal and 36S2Weri but this file
is only found in the 559 experimental directory. It appears the
lane comprising the normal retina was not sequenced until experiment 589.

The confusion likely arose because files produced in 589 have the same
names (Weri,Y79,TwoLaneWy,OneLaneWy). This coincidence is documented
in the 589 workflow document.

Using the 559 key:
\begin{verbatim}
1 FGC0031-8_GTCT
2 FGC0036-1_NOBC
3 FGC0036-2_GTCT
4 FGC0036-2_NOBC
\end{verbatim}
and workflow
\begin{verbatim}
ControlCond=1 TestCond=3 Name=Weri ;
ControlCond=2 TestCond=4 Name=Y79 ;
ControlCond=1 TestCond=2 Name=TwoLaneWy ;
ControlCond=3 TestCond=4 Name=OneLaneWy ;
\end{verbatim}
We see, for example, the document TwoLaneWy (1 vs 2) was likely a comparison of 31S8 WERI vs. 36S1 Y-79.

with following columns
\begin{enumerate}
  \item Gene - human mature miRNA
  \item ControlN and TestN - presumably non-normalized counts
  \item FoldChange and abs(FC), unknown equation presumably based on raw counts
  \item M (typically a red/green intensity ratio) definitely calculated as log$_2$ fold change
  \item A (average intensity) unknown calculation (typically (log$_2$test+log$_2$ctrl)/2)
  \item ControlP and TestP - some fraction of respective sample totals
  \item Ln(Pv) and Statistic and Heavy - unknown
\end{enumerate}

No spreadsheets were produced comparing normal with retinoblastoma samples.


\subsection{DESeq}
A reanalysis of the Schug's count data was performed using the
R-based RNA-Seq analysis package DESeq \cite{Anders} in order to
detect differential expression between normal and retinoblastoma cells.

This is a comparison of
\begin{verbatim}
\Sexpr{samp1}
and
\Sexpr{samp2}
\end{verbatim}

To deal with a lack of replicates, DESeq uses variability of
normalized gene counts with similar means, under the assumption that
most genes are not differentially expressed
(\href{http://seqanswers.com/forums/showthread.php?t=7325}{seqanswers post}). This might not be the best assumption for miRNAs.


\begin{quote}
uses the negative binomial distribution, with variance and mean linked by local regression, to model the null distribution of the count data. The method controls type-I error and provides good detection power. "
\end{quote}

Due to the lack of replicates in this data set, a "blind" method of variance estimation must be used. This approach is described by Simon Anders:
\begin{quote}
[experiments without replicates] are often undertaken, especially in HTS, and the DESeq package can deal with them, even though the soundness of the results may depend very much on the circumstances.
Our primary assumption is still that the mean is a good predictor for the variance. Hence, if a number of genes with similar expression level are compared between replicates, we expect that their variation is of comparable magnitude. Once we accept this assumption, we may argue as follows: Given two samples from different conditions and a number of genes with comparable expression levels, of which we expect only a minority to be influenced by the condition, we may take the variance estimated from comparing their count rates across conditions as ersatz for a proper estimate of the variance across replicates. After all, we assume most genes to behave the same within replicates as across conditions, and hence, the estimated variance should not change too much due to the influence of the hopefully few differentially expressed genes. Furthermore, the differentially expressed genes will only cause the variance estimate to be too high, so that the test will err to the side of being too conservative, i.e., we only lose power.
\end{quote}

<<deseqAugCounts,echo=F>>=
augCounts<-read.csv(paste(gangulyRoot,"read_counts_all_combo.mirAugCounts.csv",sep="/"),stringsAsFactors=FALSE)
augCounts<-augCounts[-1,]
normalCounts<-read.csv(paste(gangulyRoot,"all_variations_FGC0042-1_NOBC.mirAugCounts.csv",sep="/"),stringsAsFactors=FALSE)

normalCounts<-normalCounts[,c("miRNA","Counts.20.25.Upto2.unique","RPM.20.25.Upto2.unique")]
names(normalCounts)[2:3]<-c('Counts.FGC0042.1_NOBC','RPM.FGC0042.1_NOBC')
#mergedTable<-merge(normalCounts,augCounts,by.x="miRNA",by.y="miRNA")



rownames(runs)<-runs$run


#numTrueRows<-length(which(baTwoWay$Significant==TRUE))

#name agreement
miRNAfeatures<-read.table(paste(gangulyRoot,"RB_miRNA_raw_data/559-Ganguly-Chao-Solexa/miRnaFeatures.tab",sep="/"))
names(miRNAfeatures)<-c("gene","mature","x","y","variation")

#get the gene name if it is using the mature name
augCounts$mirnaGene<-laply(toupper(augCounts$miRNA),function(x){as.character(miRNAfeatures$gene[which(toupper(miRNAfeatures$gene)==x | toupper(miRNAfeatures$mature)==x)[1]])})
normalCounts$mirnaGene<-laply(toupper(normalCounts$miRNA),function(x){as.character(miRNAfeatures$gene[which(toupper(miRNAfeatures$gene)==x | toupper(miRNAfeatures$mature)==x)[1]])})

#summarized.normal<-ddply(normalCounts,.(mirnaGene),summarize,Counts.FGC0042.1_NOBC=sum(as.integer(Counts.FGC0042.1_NOBC)))
#summarized.rb<-ddply(augCounts,.(mirnaGene),summarize,Counts.FGC0031.8_GTCT=sum(as.integer(Counts.FGC0031.8_GTCT)))

evalNorm<-paste("ddply(normalCounts,.(mirnaGene),summarize,",samp2,"=sum(as.integer(",samp2,")))",sep="")
summarized.normal<-eval(parse(text=evalNorm))

evalRB<-paste("ddply(augCounts,.(mirnaGene),summarize,",samp1,"=sum(as.integer(",samp1,")))",sep="")
summarized.rb<-eval(parse(text=evalRB))

summarized<-merge(summarized.rb,summarized.normal,by="mirnaGene",all=TRUE)

@
\subsubsection{miRNA count joining}
Alignment done by Jonathan Schug were slightly different for normal
and retinoblastoma samples. The miRNA name column normal counts features a mix of mature
and precursor style nomenclatures. The retinoblastomas samples
(mirAugCounts) are strictly lowercase hsa-mir-\#\#\# style miRNA gene name entries.

To compare normal and retinoblastoma samples, mature miRNA from
normal samples were tabulated to their miRNA gene names. In some
instances this resulted in two mature miRNAs (e.g. counts of hsa-miR-122 and
hsa-miR-122* being combined to hsa-mir-122)

The normal  \begin{verbatim}\Sexpr{samp2}\end{verbatim} sample counts from \Sexpr{nrow(summarized.normal)} miRNAs
and \Sexpr{nrow(summarized.rb)} from  \begin{verbatim}\Sexpr{samp1}\end{verbatim} retinoblastoma. A full outer join of these tables produced
\Sexpr{nrow(summarized)} genes.

\subsubsection{Normalization}
DESeq can accept library sizes (referred here as the library method) to adjust for size factors, or can
estimate these sizes based on the input data (estimate method).

Another option is to use the RPM counts provided by Schug and set the
library sizes equal (RPM method)

In this report we have used the \Sexpr{useLibs} method.
<<deSetup,echo=F>>=
countsTable<-summarized[,c(samp1,samp2)]
#no nulls allowed
countsTable[is.na(countsTable)]<-1
countsTable[countsTable==0]<-1

row.names(countsTable)<-summarized$mirnaGene
conds<-as.factor(c(samp1,samp2))
cds<-newCountDataSet(countsTable,conds)

#using the read counts
if(useLibs=='library'){
sizeFactors(cds)<-runs[c(sizeName1,sizeName2),"size"]
}else{
if(useLibs=='estimate'){
#using the mirna counts
cds<-estimateSizeFactors(cds)
}else{
    if(useLibs=='rpm'){
        #sizeFactors(cds)=c(1000000,1000000)
        sizeFactors(cds)=c(1,1)
    }else{
        stop("there is something wrong with useLibs")
    }
}
}
cds<-estimateVarianceFunctions(cds,method="blind")

#always put rb in samp1
#always put control in samp2
#but call the test (s2,s1) so fold changes in rb/control
res<-nbinomTest(cds,samp2,samp1)
write.table(res,file=paste(samp1,"_",samp2,"_",useLibs,".txt",sep=""))
myPval<-.05
@

\pagebreak
\subsection{Results}
\subsubsection{Significantly expressed miRNAs}
Significantly differentially expressed miRNAs. Fold change is retinoblastoma/normal.


<<dePval,results=tex,echo=F>>=
resSig<-subset(res[,1:8],pval < myPval)
names(resSig)<-c('miRNA','baseMean','bNorm','bRB','fc','log2fc','pval','padj')
myXtableOne<-xtable(resSig[order(resSig$pval),1:8],type=tex,caption=paste("Significant (P<",myPval,") differential expression before the correction for multiple testing"),display=c('d','s','E','E','E','g','g','g','g'))

print(myXtableOne,include.rownames=FALSE, size="\\footnotesize")
@

After correcting for multiple testing (Benjamini-Hochberg procedure),
\Sexpr{nrow(subset(res,padj<.05))} miRNAs were differentially
expressed.
\pagebreak
\subsubsection{Upregulated miRNAs}
Identifying upregulated miRNAs by fold change is problematic, as the
metric does not scale well for analyses with large variations in
expression and will be dominated by low and zero count miRNAs.

One approach is to take the subset of miRNAs with baseMeans greater
than the median.

<<upreg,results=tex,echo=F>>=
regs<-subset(res[,1:8],baseMean>median(res$baseMean))
names(regs)<-c('miRNA','baseMean','bNorm','bRB','fc','log2fc','pval','padj')

upRegs<-regs[ order(-regs$fc)[1:20], ]
myXtableTwo<-xtable(upRegs,type=tex,caption=paste("Strongly upregulated miRNAs in ",gsub('_','-',samp1),sep=""),display=c('d','s','E','E','E','g','g','g','g'))
print(myXtableTwo,include.rownames=FALSE, size="\\footnotesize")
@

\pagebreak
\subsubsection{Downregulated miRNAs}
<<downreg,results=tex,echo=F>>=
downRegs<-regs[ order(regs$fc)[1:20], ]
myXtableThree<-xtable(downRegs,type=tex,caption=paste("Strongly downregulated miRNAs in ",gsub('_','-',samp1),sep=""),display=c('d','s','E','E','E','g','g','g','g'))
print(myXtableThree,include.rownames=FALSE, size="\\footnotesize")
@

\pagebreak
\subsection{Fold Change and Significance}
The method DESeq uses to estimate variance in samples with no
replicates may bias it to assigning very high fold change significance
thresholds differentially expressed genes.

<<fcVsBasemean,echo=FALSE,fig=TRUE>>=
q<-qplot(log2(res$baseMean),res$log2FoldChange)+aes(color=res$pval<.05)
print(q)
@
\pagebreak
\begin{thebibliography}{100}	% start the bibliography

\small				% put the bibliography in a small font




\bibitem{Anders} Anders, S. and  Huber, W. Differential expression analysis for sequence count data. Nature Precedings
http://dx.doi.org/10.1038/npre.2010.4282.2.

\end{thebibliography}		% end the bibliography


\end{document}
