% -*- mode: noweb; noweb-default-code-mode: R-mode; -*-
\documentclass{article}

%headers and footers
\usepackage{fancyhdr}
\setlength{\headheight}{15pt}
\pagestyle{fancyplain}
\lhead{\fancyplain{}{\thepage}}
\chead{}
\rhead{\fancyplain{}{\bfseries miRNA Reanalyis}}
\rfoot{$Rev$}
\cfoot{}
\lfoot{\includegraphics[width=0.1\textwidth]{./bicLogo.png}\\[1cm]}


%hyperlink setup
\usepackage{hyperref}
\usepackage{xcolor}
\definecolor{dark-red}{rgb}{0.4,0.15,0.15}
\definecolor{dark-blue}{rgb}{0.15,0.15,0.4}
\definecolor{medium-blue}{rgb}{0,0,0.5}
\hypersetup{
    colorlinks, linkcolor={dark-red},
    citecolor={dark-blue}, urlcolor={medium-blue}
}


\usepackage{longtable}
\usepackage{rotating}



\begin{document}
\title{A Reanalysis of miRNA differential expression, editing, and novelty}
\author{Jeremy Leipzig}
\maketitle
\tableofcontents
\pagebreak
<<setup,echo=FALSE,results=HIDE>>=
setwd("~/Documents/ganguly/trunk")
source("config.R")
library(DESeq)
library(ggplot2)
library(xtable)
library(Biostrings)

@
\section{Introduction}
miRNA counts were generated by Johnathan Schug using an Eland
alignment of miRNAs to mature miRNA sequences in mirbase and the
genome.


This analysis was revisited with the following objectives:
\begin{enumerate}
  \item Reanalyze existing miRNA count data using an existing RNA-Seq package with an explicitly documented procedure for comparing samples without replicates
  \item Use a non-heuristic aligner to reassess the existing counts
    and try the RNA-Seq analysis on these counts
  \item Explore the possibility of miRNA editing
  \item Explore potential novel miRNAs in the dataset and their
    predicted targets
\end{enumerate}

\section{Samples}
<<loadSeqs,echo=F,results=hide>>=


runs<-data.frame(run=c('FGC0042S1','FGC0031S8','FGC0036S1','FGC0036S2'),desc=c('Normal retina','WERI with barcode GTCT (Retinoblastoma)','Y-79 without barcode (Retinoblastoma)','Y-79 + WERI with barcode (Retinoblastoma)'),size=c(4787704,5821153,5877018,5897869))
@
Four Solexa runs were conducted:
\begin{enumerate}
   \item \Sexpr{paste(runs[1,1],':',runs[1,2],' ',format(runs[1,3],big.mark=",",scientific=F),' seqs',sep="")}
   \item  \Sexpr{paste(runs[2,1],':',runs[2,2],' ',format(runs[2,3],big.mark=",",scientific=F),' seqs',sep="")}
   \item \Sexpr{paste(runs[3,1],':',runs[3,2],' ',format(runs[3,3],big.mark=",",scientific=F),' seqs',sep="")}
   \item \Sexpr{paste(runs[4,1],':',runs[4,2],' ',format(runs[4,3],big.mark=",",scientific=F),' seqs',sep="")}
\end{enumerate}


\section{Realignment using Vmatch}

Eland is a first generation short read mapper with numerous
limitations.

Due to the small libraries of known miRNAs, it was possible to use a
non-heuristic general alignment tool, Vmatch, to gain total confidence
and flexibility with the distinct 11-26nt sequences collated by Schug.

\subsection{Trimmed files}
Schug had already quality trimmed, adaptor-masked, poly-A trimmed and debarcoded the
miRNA sequences from each run. In addition, sequences were split by
sequence length, presumably to focus in on true miRNA candidates and possibly
to aid Eland, which in the early days of Solexa did not accept
variable length reads.

This preprocessing comprised a decent amount of labor, and so lacking the original
adapter sequences and other minutae, so it was chosen to accept these trimmed sequences rather than start from scratch.

<<loadTrimmed,echo=F>>=
seqs<-read.table("allSeqs.txt",col.names="filePath",stringsAsFactors=FALSE)

#breakup set identifiers based on filepath, we are lucky the charlengths are all the same
seqs$seqLength<-as.integer(substr(seqs$filePath,nchar(seqs$filePath)-4,nchar(seqs$filePath)-3))
seqs$expLane<-substr(seqs$filePath,nchar(seqs$filePath)-19,nchar(seqs$filePath)-11)
seqs$barcode<-substr(seqs$filePath,nchar(seqs$filePath)-9,nchar(seqs$filePath)-6)

seqs[which(seqs$expLane=='FGC0042-1'),"sample"]<-"normal"
seqs[which(seqs$barcode=='GTCT'),"sample"]<-"WERI"
seqs[which((seqs$expLane=='FGC0036-1' | seqs$expLane=='FGC0036-2') & seqs$barcode=='NOBC'),"sample"]<-'Y-79'


#get number of sequences
seqs$numSeqs<-laply(.data=paste(gangulyRoot,seqs$filePath,sep="/"),function(x){length(read.DNAStringSet(x))})
write.table(seqs,file="schugSeqsTable.txt")
@

It was determined that no sequence within a lane/sample was repeated -
i.e. these are distinct sequences.

A breakdown of these by lane and sample

<<seqTab,echo=F>>=
#seqTabulation<-format(ddply(seqs,c("sample","expLane"),summarize,distinctSeqs=sum(numSeqs)),big.mark=",")
#myXtableTwo<-xtable(seqTabulation,type=tex,caption=paste("Distinct sequences by sample/lane"))
#print(myXtableTwo,include.rownames=FALSE, size="\\footnotesize")

rbDir<-"../RB_miRNA_raw_data/559-Ganguly-Chao-Solexa/MaskAndSplit/Combo/Splits/"
rbLines<-c("FGC0031-8_GTCT","FGC0036-2_NOBC","FGC0036-1_NOBC","FGC0036-2_GTCT")
cols<-c("seqs","instances","seqLength")
rbTab<-ldply(.data=rbLines,.fun=function(x){cbind(sample=x,read.table(paste(rbDir,x,"_report.tab",sep=""),col.names=cols))})


normTab<-cbind(sample="FGC0042-1_NOBC",read.table("../RB_miRNA_raw_data/589-Ganguly-Chao-Solexa_smRNA/MaskAndSplit/Splits/FGC0042-1_NOBC_report.tab",col.names=cols))
allTab<-rbind(rbTab,normTab)
allTab$sample<-as.character(allTab$sample)
allTab$sample<-gsub(allTab$sample,pattern='_',replacement='-')
allTab$sample<-as.factor(allTab$sample)
seqTab<-allTab
caption<-"All sequences by sample"
@
<<printTab,results=tex,echo=F>>=
seqTabulation<-format(ddply(seqTab,"sample",summarize,distinctSeqs=length(seqs),weightedSeqs=sum(instances)),big.mark=",")
myXtableTwo<-xtable(seqTabulation,type=tex,caption=caption)
print(myXtableTwo,include.rownames=FALSE, size="\\footnotesize")
@

<<sublenSetup,echo=F>>=
minLength<-21
maxLength<-26
seqTab<-subset(allTab,seqLength>=minLength & seqLength<=maxLength)
caption<-paste("Sequences ",minLength,"-",maxLength,sep="")
@
The subset of sequences with lengths in the range \Sexpr{minLength}-\Sexpr{maxLength}nt inclusive is as follows:
<<seqTab2,results=tex,echo=F>>=
<<printTab>>
@


<<sampleSetup,echo=F>>=
samp1<-'normal'
file1<-'normals'
vmfile1<-'normals'
expLane1<-'FGC0042-1'
barcode1<-'NOBC'
samp2<-'WERI36'
file2<-'WERI_36'
vmfile2<-'weri36'
expLane2<-'FGC0036-2'
barcode2<-'GTCT'

attach(seqs)
sampsize1<-sum(seqs[which(seqLength >= minLength & seqLength <= maxLength & sample==samp1 &  expLane==expLane1),"numSeqs"])
sampsize2<-sum(seqs[which(seqLength >= minLength & seqLength <= maxLength & sample==samp2 &  expLane==expLane2),"numSeqs"])
hairpins<-readFASTA("../mirna/hairpin.fa",strip.descs=TRUE)
@

\subsection{Against hairpin miRNAs}
For the purposes of editing, we can examine how many of the sequences between \Sexpr{minLength}-\Sexpr{maxLength}nt align with zero, one, or two edits.

These are aligned to the subset of human miRNAs in hairpin.fa (\Sexpr{format(length(hairpins),big.mark=",",scientific=F)})
An edit is defined as an insertion, deletion or substitution event.

In this report we will compare

\textbf{\Sexpr{samp1} \Sexpr{expLane1}}:\Sexpr{format(sampsize1,big.mark=",",scientific=F)} \Sexpr{minLength}-\Sexpr{maxLength}nt seqs

and

\textbf{\Sexpr{samp2} \Sexpr{expLane2}}:\Sexpr{format(sampsize2,big.mark=",",scientific=F)} \Sexpr{minLength}-\Sexpr{maxLength}nt seqs

Alignment is against the forward strand only, reverse complements are not considered.
<<makeFasta,echo=F>>=

#write.XStringSet(read.DNAStringSet(paste(gangulyRoot,filePath[which(seqLength >= 21 & seqLength <=26 & sample=='normal' &  expLane=='FGC0042-1')],sep="/")),file="normals.fa",format="fasta")
#write.XStringSet(read.DNAStringSet(paste(gangulyRoot,filePath[which(seqLength >= 21 & seqLength <=26 & sample=='WERI'   &  expLane=='FGC0036-2')],sep="/")),file="WERI_36.fa",format="fasta")
@
\subsection{Counting principles}
Counting is query sequence centric, rather than miRNA centric.
Sequences that align exactly to any miRNA and not considered for editing.
Sequences that align to multiple miRNAs are tallied only once, this rule extends to sequences that align with one edit to any miRNA and to different miRNA with two edits.

vmatch  -showdesc 30  -complete -e 2


<<readVmatch,echo=F>>=
vmnames<-c("tLen","tName","tPos","strand","qLen","qName","qPos","dist","e-value","score","identity")
#res1.0<-read.table(paste(vmfile1,"exact.vmatch",sep="."),col.names=vmnames)
#res2.0<-read.table(paste(vmfile2,"exact.vmatch",sep="."),col.names=vmnames)

#one mismatch -we don't want exact
#res1.1<-read.table(paste(vmfile1,"e1.all.vmatch",sep="."),col.names=vmnames)
#res1.1<-res1.1[-which(res1.1$qName %in% res1.0$qName),]
#res2.1<-read.table(paste(vmfile2,"e1.all.vmatch",sep="."),col.names=vmnames)
#res2.1<-res2.1[-which(res2.1$qName %in% res2.0$qName),]

res1.2<-read.table(paste(vmfile1,"e2.all.vmatch",sep="."),col.names=vmnames)
tab1<-subset(allTab,sample==paste(expLane1,barcode1,sep='-'))
res1.2nr<-res1.2[-which(res1.2$qName %in% res1.2[which(res1.2$dist==1),"qName"] & res1.2$dist==2),]
exact1<-res1.2nr[which(res1.2nr$dist==0),]
oneEdit1<-res1.2nr[which(res1.2nr$dist==1),]
twoEdit1<-res1.2nr[which(res1.2nr$dist==2),]


res2.2<-read.table(paste(vmfile2,"e2.all.vmatch",sep="."),col.names=vmnames)
tab2<-subset(allTab,sample==paste(expLane2,barcode2,sep='-'))
res2.2nr<-res2.2[-which(res2.2$qName %in% res2.2[which(res2.2$dist==1),"qName"] & res2.2$dist==2),]
exact2<-res2.2nr[which(res2.2nr$dist==0),]
oneEdit2<-res2.2nr[which(res2.2nr$dist==1),]
twoEdit2<-res2.2nr[which(res2.2nr$dist==2),]

merged2<-merge(exact2,tab2,by.x="qName",by.y="seqs",all.x=FALSE)

sumCnts<-data.frame(sample=c(samp1,samp2),
  seqs=c(nrow(tab1),nrow(tab2)),
  noeditAlign=c(nrow(exact1),nrow(exact2)),
  noeditDS=c(length(levels(factor(exact1$qName))),length(levels(factor(exact2$qName)))),
  noeditMR=c(length(levels(factor(exact1$tName))),length(levels(factor(exact2$tName)))),
  oneeditAlign=c(nrow(oneEdit1),nrow(oneEdit2)),
  oneeditDS=c(length(levels(factor(oneEdit1$qName))),length(levels(factor(oneEdit2$qName)))),
  oneeditMR=c(length(levels(factor(oneEdit1$tName))),length(levels(factor(oneEdit2$tName)))),
  twoeditAlign=c(nrow(twoEdit1),nrow(twoEdit2)),
  twoeditDS=c(length(levels(factor(twoEdit1$qName))),length(levels(factor(twoEdit2$qName)))),
  twoeditMR=c(length(levels(factor(twoEdit1$tName))),length(levels(factor(twoEdit2$tName))))
  )

#edited1<-res1.2[!(res1.2$qName %in% res1.2[which(res1.2$dist==1),"qName"]),]
#res1.2<-res1.2[-which(res1.2$qName %in% res1.0$qName),]
#res1.2<-res1.2[-which(res1.2$qName %in% res1.1$qName),]


#twoEditHuh<-res2.2[!(res2.2$qName %in% res2.2[which(res2.2$dist==1),"qName"]),]

#res2.2<-res2.2[-which(res2.2$qName %in% res2.0$qName),]
#res2.2<-res2.2[-which(res2.2$qName %in% res2.1$qName),]


#qplot(
#f <- function(s) strsplit(as.character(s), "_")[[1]][1]
#sapply(res2.2$tName, f)


#    V1                        V2 V3 V4 V5                        V6 V7 V8       V9 V10    V11
#1   25 hsa-mir-181a-1_MI0000289_ 23  D 25 AAAATTCAACGCTGTCGGTGAGTTT  0  1 9.48e-09  47  96.00
@

\begin{thebibliography}{100}	% start the bibliography

\small				% put the bibliography in a small font

\bibitem{Anders} Anders, S. and  Huber, W. Differential expression analysis for sequence count data. Nature Precedings
http://dx.doi.org/10.1038/npre.2010.4282.2.

\end{thebibliography}		% end the bibliography


\end{document}
