% -*- mode: noweb; noweb-default-code-mode: R-mode; -*-
\documentclass{article}

%headers and footers
\usepackage{fancyhdr}
\setlength{\headheight}{15pt}
\pagestyle{fancyplain}
\lhead{\fancyplain{}{\thepage}}
\chead{}
\rhead{\fancyplain{}{\bfseries miRNA HiSeq Analysis}}
\rfoot{$Rev: 2964 $}
\cfoot{}
\lfoot{\includegraphics[width=0.1\textwidth]{./bicLogo.png}\\[1cm]}

\usepackage[margin=1in]{geometry}

%hyperlink setup
\usepackage{hyperref}
\usepackage{xcolor}
\definecolor{dark-red}{rgb}{0.4,0.15,0.15}
\definecolor{dark-blue}{rgb}{0.15,0.15,0.4}
\definecolor{medium-blue}{rgb}{0,0,0.5}
\hypersetup{
    colorlinks, linkcolor={dark-red},
    citecolor={dark-blue}, urlcolor={medium-blue}
}


\usepackage{longtable}
\usepackage{rotating}

%underscores in variable names cause problems
%\usepackage{underscore}

\begin{document}
\newcommand{\COT}{C$_{0}$t }
\title{An analysis of miRNA differential expression from HiSeq data (bwa alignment)}
\author{Jeremy Leipzig}
\maketitle
\tableofcontents
\pagebreak
<<setup,echo=FALSE,results=hide>>=
library(Rsamtools)
library(Biostrings)
library(stringr)
library(plyr)
library(reshape)
#library("edgeR")
library("DESeq")
library(ggplot2)
library(xtable)

rpmNormalize<-FALSE
minCount<-100
s <- read.DNAStringSet("refs/hsa_hairpin.dna.fa",use.names=TRUE)
hsa<-s[str_detect(names(s),"^hsa")]

names(hsa)<-str_match(names(hsa),"^\\S+")
hsaGR<-GRanges(names(hsa),IRanges(1,width(hsa)),strand="*")


mirna<-function(x){
  xdf<-countBam(file=paste(bamDirectory,x,suffix,sep=""),param=ScanBamParam(which=hsaGR))[,c("space","records")]
  names(xdf)<-c("miRNA","reads")
  if(rpmNormalize){xdf$reads<-xdf$reads*1000000/sum(xdf$reads)}
  xdf$variable<-x
  xdf
}

bamDirectory<-"bwaAlignments/"
mismatches<-1
#"./"
conds<-c("N","T","N","T","N","T")
samples<-c("RB494N","RB494T","RB495N","RB495T","RB498N","RB498T")
suffix<-paste(".fq.cutadapt.min17.hairpin",mismatches,"sorted.bam",sep=".")
#".hairpin.bam"


uncasted<-ldply(as.list(samples),.fun=mirna)
rnaSeqReadyDataFrame<-cast(uncasted,miRNA ~ variable,value="reads")
write.csv(rnaSeqReadyDataFrame,file="rawCounts.csv")
rnaSeqReadyDataFrame<-subset(rnaSeqReadyDataFrame,rowSums(rnaSeqReadyDataFrame)>minCount)

countsTable<-as.matrix(rnaSeqReadyDataFrame[,-1])
row.names(countsTable)<-rnaSeqReadyDataFrame$miRNA



#d<-DGEList(counts=countsTable,group=conds)
#d<-calcNormFactors(d)
#d<-estimateCommonDisp(d)
#de.com <- exactTest(d)
#topTags(de.com)





cds <- newCountDataSet( countsTable, conds )
cds <- estimateSizeFactors( cds )
cds <- estimateVarianceFunctions( cds )
res <- nbinomTest( cds, "N", "T")
write.csv(res,file="deseqResults.csv")

@
\section{Introduction}
Six samples, three cohorts of normal and retinoblastoma tissues were collected for miRNA sequencing using the Illumina HiSeq platform.
miRNA counts were generated by a BWA alignment to hairpin sequences without correcting for potential duplicates.



The alignment allows \Sexpr{mismatches} mismatches.



BWA assigns multiple hits randomly. Multiple members of a miRNA family can inheriently create multiple hits, so the results here should be viewed carefully.

\begin{verbatim}
export samples="RB494N.fq RB494T.fq RB495N.fq RB495T.fq RB498N.fq RB498T.fq"
export settings=".cutadapt.min17"
export ref="hairpin"

parallel cutadapt -a ATCTCGTATGCCGTCTTCTGCTTG {} -m 17 ">" {}"$settings" ::: $samples
                                                                                                                                                                                                                
parallel bwa aln -n \Sexpr{mismatches} -I $PWD/refs/$ref cutadapt/{}"$settings" 
  ">" sai/{}"$settings.$ref.sai" ::: $samples \
&& parallel bwa samse -n 1 $PWD/refs/$ref sai/{}"$settings.$ref.sai" cutadapt/{}"$settings" 
  ">" bwaAlignments/{}"$settings.$ref.sam" ::: $samples \
&& parallel samtools view -b -S bwaAlignments/{}"$settings.$ref.sam" 
  ">" bwaAlignments/{}"$settings.$ref.bam" ::: $samples \
&& parallel --jobs 3 samtools sort bwaAlignments/{}"$settings.$ref.bam" 
  bwaAlignments/{}"$settings.$ref.sorted" ::: $samples     
\end{verbatim}

\subsection{DESeq}
A reanalysis of the non-normalized count data was performed using the
R-based RNA-Seq analysis package DESeq \cite{Anders} in order to
detect differential expression between normal and retinoblastoma cells.

Only miRNAs with a combined count of \Sexpr{minCount} among all samples were considered.
\Sexpr{nrow(rnaSeqReadyDataFrame)} miRNAs fit this criteria.

\pagebreak
\subsection{Results}
\subsubsection{Significantly expressed miRNAs}
Significantly differentially expressed miRNAs. Fold change is retinoblastoma/normal.

<<dePval,results=tex,echo=F>>=
myPval<-.05
resSig<-subset(res[,1:8],pval < myPval)
names(resSig)<-c('miRNA','baseMean','bNorm','bRB','fc','log2fc','pval','padj')
myXtableOne<-xtable(resSig[order(resSig$pval),1:8],type=tex,caption=paste("Significant (P<",myPval,") differential expression before the correction for multiple testing"),display=c('d','s','E','E','E','g','g','g','g'))

print(myXtableOne,include.rownames=FALSE, size="\\footnotesize")
@

\pagebreak
\subsubsection{Upregulated miRNAs}
<<upreg,results=tex,echo=F>>=
#upRegs<-resSig[ order(-resSig$fc)[1:20], ]
upRegs<-head(subset(resSig[order(-resSig$fc),],fc>1),n=20)


myXtableTwo<-xtable(upRegs,type=tex,caption=paste("Strongly significantly upregulated miRNAs in Retinoblastoma",sep=""),display=c('d','s','E','E','E','g','g','g','g'))
print(myXtableTwo,include.rownames=FALSE, size="\\footnotesize")
@

\pagebreak
\subsubsection{Downregulated miRNAs}
<<downreg,results=tex,echo=F>>=
downRegs<-head(subset(resSig[order(resSig$fc),],fc<=1),n=20)


myXtableThree<-xtable(downRegs,type=tex,caption=paste("Strongly significantly downregulated miRNAs in Retinoblastoma",sep=""),digits=3,display=c('d','s','E','E','E','g','g','g','g'))
print(myXtableThree,include.rownames=TRUE, size="\\footnotesize")
@

\pagebreak
\subsubsection{hsa-let miRNAs}
<<downreg,results=tex,echo=F>>=
letRegs<-res[substr(res$id,0,7)=='hsa-let',1:8]
names(letRegs)<-c('miRNA','baseMean','bNorm','bRB','fc','log2fc','pval','padj')
myXtableFour<-xtable(letRegs,type=tex,caption=paste("hsa-let miRNAs in Retinoblastoma",sep=""),digits=3,display=c('d','s','E','E','E','g','g','g','g'))
print(myXtableFour,include.rownames=TRUE, size="\\footnotesize")
@

\pagebreak
\subsection{Fold Change and Significance}
Genes with high fold change may not be significantly differentially expressed simply due to high variance.

<<fcVsBasemean,echo=FALSE,fig=TRUE>>=
q<-qplot(log2(res$baseMean),res$log2FoldChange)+aes(color=res$pval<.05)
print(q)
@


\begin{thebibliography}{100}  % start the bibliography

\small				% put the bibliography in a small font


\bibitem{Anders} Anders, S. and  Huber, W. Differential expression analysis for sequence count data. Nature Precedings
http://dx.doi.org/10.1038/npre.2010.4282.2.

\end{thebibliography}  	% end the bibliography


\end{document}

