% -*- mode: noweb; noweb-default-code-mode: R-mode; -*-
\documentclass{article}

%headers and footers
\usepackage{fancyhdr}
\setlength{\headheight}{15pt}
\pagestyle{fancyplain}
\lhead{\fancyplain{}{\thepage}}
\chead{}
\rhead{\fancyplain{}{\bfseries miRNA Conkrite Analysis}}
\rfoot{$Rev: 2964 $}
\cfoot{}
\lfoot{\includegraphics[width=0.1\textwidth]{./bicLogo.png}\\[1cm]}

\usepackage[margin=1in]{geometry}

%hyperlink setup
\usepackage{hyperref}
\usepackage{xcolor}

\definecolor{dark-red}{rgb}{0.4,0.15,0.15}
\definecolor{dark-blue}{rgb}{0.15,0.15,0.4}
\definecolor{medium-blue}{rgb}{0,0,0.5}
\hypersetup{
    colorlinks, linkcolor={dark-red},
    citecolor={dark-blue}, urlcolor={medium-blue}
}


\usepackage{longtable}
\usepackage{rotating}

%underscores in variable names cause problems
%\usepackage{underscore}

\begin{document}
\newcommand{\COT}{C$_{0}$t }
\title{An analysis of miRNA differential expression from Conkrite data}
\author{Jeremy Leipzig}
\maketitle
\tableofcontents
\pagebreak
<<setup,echo=FALSE,results=hide>>=
library(Rsamtools)
library(Biostrings)
library(stringr)
library(plyr)
library(reshape)
#library("edgeR")
library("DESeq")
library(ggplot2)
library(xtable)

rpmNormalize<-FALSE
minCount<-200

workDir<-"/nas/is1/leipzig/Ganguly/gangulyRBhi/"
counts<-read.table(paste(workDir,"data/conkrite/conkrite.txt",sep=""),header=TRUE)

conds<-c("T","T","N","N")
samples<-c("rb-1","rb-2","fw18-1","fw18-2")

aboveThresholdCounts<-subset(counts,rowSums(counts[-1])>minCount)

countsTable<-as.matrix(aboveThresholdCounts[,-1])
row.names(countsTable)<-as.character(aboveThresholdCounts$microRNA)





cds <- newCountDataSet( countsTable, conds )
cds <- estimateSizeFactors( cds )
cds <- estimateVarianceFunctions( cds )
res <- nbinomTest( cds, "N", "T")
write.csv(res,file="deseqResults.csv")

@
\section{Introduction}
This is an DESeq differential expression analysis of raw, non-unique counts reported as supplementary material from the 
\cite{Conkrite} paper on mir-17-92 family.



\subsection{DESeq}
A reanalysis of the non-normalized count data was performed using the
R-based RNA-Seq analysis package DESeq \cite{Anders} in order to
detect differential expression between normal and retinoblastoma cells.

Only miRNAs with a combined count of \Sexpr{minCount} among all samples were considered.
\Sexpr{nrow(aboveThresholdCounts)} miRNAs fit this criteria.

DESeq performs its own normalization routine so raw counts are preferable as input.

\pagebreak
\subsection{Results}
\subsubsection{Significantly expressed miRNAs}
Significantly differentially expressed miRNAs. Fold change is retinoblastoma/normal.

<<dePval,results=tex,echo=F>>=
myPval<-.05
resSig<-subset(res[,1:8],pval < myPval)
names(resSig)<-c('miRNA','baseMean','bNorm','bRB','fc','log2fc','pval','padj')
myXtableOne<-xtable(resSig[order(resSig$pval),1:8],type=tex,caption=paste("Significant (P<",myPval,") differential expression before the correction for multiple testing"),display=c('d','s','E','E','E','g','g','g','g'))

print(myXtableOne,include.rownames=FALSE, tabular.environment = "longtable", floating = FALSE, size="\\footnotesize")
@

\pagebreak
\subsubsection{Upregulated miRNAs}
<<upreg,results=tex,echo=F>>=
#upRegs<-resSig[ order(-resSig$fc)[1:20], ]
upRegs<-head(subset(resSig[order(-resSig$fc),],fc>1),n=20)


myXtableTwo<-xtable(upRegs,type=tex,caption=paste("Strongly significantly upregulated miRNAs in Retinoblastoma",sep=""),display=c('d','s','E','E','E','g','g','g','g'))
print(myXtableTwo,include.rownames=FALSE, size="\\footnotesize")
@

\pagebreak
\subsubsection{Downregulated miRNAs}
<<downreg,results=tex,echo=F>>=
downRegs<-head(subset(resSig[order(resSig$fc),],fc<=1),n=20)


myXtableThree<-xtable(downRegs,type=tex,caption=paste("Strongly significantly downregulated miRNAs in Retinoblastoma",sep=""),digits=3,display=c('d','s','E','E','E','g','g','g','g'))
print(myXtableThree,include.rownames=FALSE, size="\\footnotesize")
@

\pagebreak
\subsubsection{hsa-let miRNAs}
<<letreg,results=tex,echo=F>>=
letRegs<-res[substr(res$id,0,7)=='hsa-let',1:8]
names(letRegs)<-c('miRNA','baseMean','bNorm','bRB','fc','log2fc','pval','padj')
myXtableFour<-xtable(letRegs,type=tex,caption=paste("hsa-let miRNAs in Retinoblastoma",sep=""),digits=3,display=c('d','s','E','E','E','g','g','g','g'))
print(myXtableFour,include.rownames=FALSE, size="\\footnotesize")
@

\pagebreak
\subsubsection{miR-17~92 cluster miRNAs}
<<clusterreg,results=tex,echo=F>>=
clusterMembers<-c('hsa-miR-20a','hsa-miR-92a','hsa-miR-93','hsa-miR-106b','hsa-miR-17')
clusterRegs<-res[res$id %in% clusterMembers,1:8]
names(clusterRegs)<-c('miRNA','baseMean','bNorm','bRB','fc','log2fc','pval','padj')
myXtableFive<-xtable(clusterRegs,type=tex,caption=paste("miR-17~92 cluster miRNAs in Retinoblastoma",sep=""),digits=3,display=c('d','s','E','E','E','g','g','g','g'))
print(myXtableFive,include.rownames=FALSE, size="\\footnotesize")
@
\pagebreak
\subsection{Fold Change and Significance}
Genes with high fold change may not be significantly differentially expressed simply due to high variance.

<<fcVsBasemean,echo=FALSE,fig=TRUE>>=
q<-qplot(log2(res$baseMean),res$log2FoldChange)+aes(color=res$pval<.05)
print(q)
@


\begin{thebibliography}{100}  % start the bibliography

\small				% put the bibliography in a small font

\bibitem{Conkrite} Conkrite, K et al. miR-17~92 cooperates with RB pathway mutations to promote retinoblastoma. Genes Dev. 2011 25: 1734-1745. doi:10.1101/gad.17027411

\bibitem{Anders} Anders, S. and  Huber, W. Differential expression analysis for sequence count data. Nature Precedings
http://dx.doi.org/10.1038/npre.2010.4282.2.

\end{thebibliography}  	% end the bibliography


\end{document}

